### ISSUE TITLE
Unhandled exception when trying to fill the library

### ISSUE THREAD

**Issue by GaMeRaM** [JMTK/SunshineGameFinder#29]
I did run it as administrator.
It doesn't seem to be based on the specific game. 
When i moved Abiotic Factor out of the folder - it failed on the next game that had exe in its folder  
![image](https://github.com/user-attachments/assets/14f7918b-d074-4eb1-9f0f-d104728f0fb0)

**Comment by JMTK** [comment]
> ###
> **[GaMeRaM](/GaMeRaM) ** commented [Nov 8, 2024](#issue-2644780223)

I just pushed a new version, 1.8.2, can you try it out?
https://github.com/JMTK/SunshineGameFinder/releases/tag/1.8.2

**Comment by GaMeRaM** [comment]
I appreciate the fast fix!

Seems like there is another problem. Sunshine on windows wraps booleans and numbers in quotes. And it breaks json parser in SunshineGameFinder:  
![image](https://github.com/user-attachments/assets/91227610-4c04-487c-8809-c83207422b6f)  
![image](https://github.com/user-attachments/assets/254bc237-a3a2-4390-879f-ad7417b37457)
When i convert them from strings to the actual types GameFinder works correctly, but resulting config seems to break the Sunshine. No games are shown in Moonlight on android  

There is also a difference that Sunshine writes empty parameters like `output` as empty strings but GameFinder writes them as `null`

Also there is a note in Sunshine UI that if path to executable contains spaces - you should wrap the path in quotes. GameFinder currently doesnt do that (and kinda breaks previous games. It now added unicode characters instead of escaped quotes)
![image](https://github.com/user-attachments/assets/1b8c88a2-20cc-4869-80a8-9736cc8c97e2)

**Comment by JMTK** [comment]
> I appreciate the fast fix!
> 
> Seems like there is another problem. Sunshine on windows wraps booleans and numbers in quotes. And it breaks json parser in SunshineGameFinder: ![image](https://private-user-images.githubusercontent.com/2902911/384875237-91227610-4c04-487c-8809-c83207422b6f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzc3ODIsIm5iZiI6MTczMTMzNzQ4MiwicGF0aCI6Ii8yOTAyOTExLzM4NDg3NTIzNy05MTIyNzYxMC00YzA0LTQ4N2MtODgwOS1jODMyMDc0MjJiNmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTFUMTUwNDQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTZhODQ4NzljNzBiNGI0NDcwM2I3OTc4YmY0NThkMGIzODU2ZTBkZmYwZmVmNTFiMDAxYTdlODllOTM0MTIxMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.JDItj1CCt1gWydWVT_gsXuRN9tRc85AhqwVX9-KCj2U) ![image](https://private-user-images.githubusercontent.com/2902911/384875308-254bc237-a3a2-4390-879f-ad7417b37457.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzc3ODIsIm5iZiI6MTczMTMzNzQ4MiwicGF0aCI6Ii8yOTAyOTExLzM4NDg3NTMwOC0yNTRiYzIzNy1hM2EyLTQzOTAtODc5Zi1hZDc0MTdiMzc0NTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTFUMTUwNDQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Nzc3ZTJkNzY4NjYxOTJmMGI4MGNlMjAzOWQxNGYxNzRjYTI4YzQwOTg4ZTRjZTYzZDE0ZWNlNjNlZDU3ZWQzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Ynl2teb-bG0O7yQIGUaz-ph9OPDfpBDYWSGVUNob1S4) When i convert them from strings to the actual types GameFinder works correctly, but resulting config seems to break the Sunshine. No games are shown in Moonlight on android
> 
> There is also a difference that Sunshine writes empty parameters like `output` as empty strings but GameFinder writes them as `null`
> 
> Also there is a note in Sunshine UI that if path to executable contains spaces - you should wrap the path in quotes. GameFinder currently doesnt do that (and kinda breaks previous games. It now added unicode characters instead of escaped quotes) ![image](https://private-user-images.githubusercontent.com/2902911/384876523-1b8c88a2-20cc-4869-80a8-9736cc8c97e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzc3ODIsIm5iZiI6MTczMTMzNzQ4MiwicGF0aCI6Ii8yOTAyOTExLzM4NDg3NjUyMy0xYjhjODhhMi0yMGNjLTQ4NjktODBhOC05NzM2Y2M4Yzk3ZTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTFUMTUwNDQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDUxODFiMGFkMjU0Mjk4MzQ0NGQ4MmMzNzE1M2U1OWJiYTVhOGNlMzFkMTIxNDZmYjYxODIwZGQwNjZmZTczNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zm5Y8cpK5u0K7RXFOP5sUZfbuKCprMTtduKfXU48hTI)

Would you mind sharing your apps.json so I can use it for testing? It seems your configuration varies wildly from what I am testing against so I wanna make sure I meet all the use cases

**Comment by JMTK** [comment]
Hi @GaMeRaM I was able to reproduce your issue, would you kindly try 1.9.1? https://github.com/JMTK/SunshineGameFinder/releases/tag/1.9.1-0

If that's still not working, would you be able to share your apps.json so I can further troubleshoot?

Thank you!

**Comment by GaMeRaM** [comment]
> Hi @GaMeRaM I was able to reproduce your issue, would you kindly try 1.9.2? https://github.com/JMTK/SunshineGameFinder/releases/tag/1.9.2
> 
> If that's still not working, would you be able to share your apps.json so I can further troubleshoot?
> 
> Thank you!

I'm sorry, but seems like this release doesnt contain a compiled binary

Added three files:
1. originally generated by sunshine and one game added by hand: [1_apps_sunshine_generated.json](https://github.com/user-attachments/files/17714367/1_apps_sunshine_generated.json)
2. original file fixed by hand to make it work with gamefinder: [2_apps_sunshine_compatible.json](https://github.com/user-attachments/files/17714369/2_apps_sunshine_compatible.json)
3. apps.json generated by GameFinder based on File 2: [3_apps_gamefinder.json](https://github.com/user-attachments/files/17714370/3_apps_gamefinder.json)

**Comment by JMTK** [comment]
> > Hi @GaMeRaM I was able to reproduce your issue, would you kindly try 1.9.2? https://github.com/JMTK/SunshineGameFinder/releases/tag/1.9.2
> > If that's still not working, would you be able to share your apps.json so I can further troubleshoot?
> > Thank you!
> 
> I'm sorry, but seems like this release doesnt contain a compiled binary
> 
> Added three files:
> 
> 1. originally generated by sunshine and one game added using sunshine web ui: [1_apps_sunshine_generated.json](https://github.com/user-attachments/files/17714367/1_apps_sunshine_generated.json)
> 2. original file fixed by hand to make it work with gamefinder: [2_apps_sunshine_compatible.json](https://github.com/user-attachments/files/17714369/2_apps_sunshine_compatible.json)
> 3. apps.json generated by GameFinder based on File 2: [3_apps_gamefinder.json](https://github.com/user-attachments/files/17714370/3_apps_gamefinder.json)

Weird, sorry about that. Seems like the upload originally failed. Should be good now. I'll test out your files in the meantime! Thank you

**Comment by JMTK** [comment]
> Tried it with file 1
> 
> ![image](https://github.com/user-attachments/assets/a0ae6c18-156f-4641-8fe0-84b9fa04504d)
> 
> Seems like it doesnt like empty env key in apps.json (which comes empty default with sunshine)

Interesting, mine is also empty but I'm not seeing that issue. I'll do some more testing!

**Comment by GaMeRaM** [comment]
Tried to fix it myself and it seems to be working correctly
https://github.com/GaMeRaM/SunshineGameFinder/tree/json-serialiser-fix
(Sorry if code looks jank. I'm not a C# programmer, so there is some AI stuff that may be weird)

Also found a bug. GameFinder will remove any game that starts through 'steam://...' cmd if removeUninstalled is true

**Comment by JMTK** [comment]
> Tried to fix it myself and it seems to be working correctly (however it may be wrong in Env departament) https://github.com/GaMeRaM/SunshineGameFinder/tree/json-serialiser-fix (Sorry if code looks jank. I'm not a C# programmer, so there is some AI stuff that may be weird)
> 
> Also found a bug. GameFinder will remove any game that starts through 'steam://...' cmd if removeUninstalled is true

Awesome I think your code looks good! Thanks for looking into it. Feel free to go ahead and open a PR and list all the improvements I'd be happy to merge it in!

**Comment by GaMeRaM** [comment]
https://github.com/JMTK/SunshineGameFinder/pull/33

**Comment by strazto** [comment]
Sounds like the solution is in the pipeline but I get a similar issue with 1.9.2 , related to the env key it seems:

```
Unhandled exception: Newtonsoft.Json.JsonSerializationException: Unable to find a constructor to use for type SunshineGameFinder.SunshineConfig. A class should either have a default constructor, one constructor with arguments or a constructor marked with the JsonConstructor attribute. Path 'env', line 2, position 10.
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader, JsonObjectContract, JsonProperty , JsonProperty , String , Boolean& )
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader , Type, JsonContract, JsonProperty, JsonContainerContract, JsonProperty, Object)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader , Type, JsonContract, JsonProperty, JsonContainerContract, JsonProperty, Object)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader , Type, Boolean)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader , Type)
   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader , Type)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String , Type, JsonSerializerSettings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String , JsonSerializerSettings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String )
   at Program.<>c__DisplayClass0_0.<<Main>$>b__0(String[] addlDirectories, String[] addlExeExclusionWords, String sunshineConfigLocation, Boolean forceUpdate, Boolean removeUninstalled)
   at System.CommandLine.Handler.<>c__DisplayClass6_0`5.<SetHandler>b__0(InvocationContext context)
   at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

**Comment by GaMeRaM** [comment]
> Sounds like the solution is in the pipeline but I get a similar issue with 1.9.2 , related to the env key it seems:
> 
> ```
> Unhandled exception: Newtonsoft.Json.JsonSerializationException: Unable to find a constructor to use for type SunshineGameFinder.SunshineConfig. A class should either have a default constructor, one constructor with arguments or a constructor marked with the JsonConstructor attribute. Path 'env', line 2, position 10.
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader, JsonObjectContract, JsonProperty , JsonProperty , String , Boolean& )
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader , Type, JsonContract, JsonProperty, JsonContainerContract, JsonProperty, Object)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader , Type, JsonContract, JsonProperty, JsonContainerContract, JsonProperty, Object)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader , Type, Boolean)
>    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader , Type)
>    at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader , Type)
>    at Newtonsoft.Json.JsonConvert.DeserializeObject(String , Type, JsonSerializerSettings)
>    at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String , JsonSerializerSettings)
>    at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String )
>    at Program.<>c__DisplayClass0_0.<<Main>$>b__0(String[] addlDirectories, String[] addlExeExclusionWords, String sunshineConfigLocation, Boolean forceUpdate, Boolean removeUninstalled)
>    at System.CommandLine.Handler.<>c__DisplayClass6_0`5.<SetHandler>b__0(InvocationContext context)
>    at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext)
>    at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
> --- End of stack trace from previous location ---
>    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
> ```

Can you please send your apps.json so I can test the updated code?

**Comment by JMTK** [comment]
Fixed in 1.10.0
